File: swap-nodes-in-pairs/swap-nodes-in-pairs.py
"""
time: O(len(linked-list)) ~ O(n)
space: O(n)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# ex input: 1 -> 2 -> 3 -> nil 
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        tmp = self.swapPairs(head.next.next) 
        newHead = head.next 
        newHead.next = head
        head.next = tmp
        return newHead

File: swap-nodes-in-pairs/swap-nodes-in-pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func swapPairs(head *ListNode) *ListNode {  
    // len == 1
    if head == nil || head.Next == nil {
        return head
    }
    headPtr := head.Next //for return
    n1 := head 
    //currTail := n1
    n2 := head.Next 
    n3 := n2.Next 
    var currTail *ListNode 
    for {
        n2.Next = n1
        n1.Next = n3
        currTail = n1 
        if n3 == nil || n3.Next == nil {
            break
        }
        n1 = n3 
        n2 = n3.Next
        n3 = n2.Next
        currTail.Next = n2
        
    }
    return headPtr
    
}

